bolt-simple-taproot.md

```diff
@@ -0,0 +1,1325 @@
+```
+
+In the case of a routine force close, the script path must be revealed so the
+broadcaster can sweep their funds after a delay. The control block to spend is
+only `33` bytes, as it just includes the internal key (along with the y-parity
+bit and leaf version):
+``` 
+delay_control_block = (output_key_y_parity | 0xc0) || taproot_nums-point || to_delay_srcipt
+```
+
+A valid witness is then:
+```
+<local_delayedsig> <to_delay_script> <delay_control_block>
+```
+
+As with base channels, the `nSequence` field must be set to `to_self_delay`.
+
+#### To Remote Outputs
+
+As we inherit the anchor output, semantics we want to ensure that the remote
+party can unilaterally sweep their funds after the 1 block CSV delay. In order
+to achieve this property, we'll utilize a NUMS point (`simple_taproot_nums`) By
+using this point as the internal key, we ensure that the remote party isn't
+able to by pass the CSV delay.
+
+Using a NUMS key has a key benefit: the static internal key allows the remote
+party to scan for their output on chain, which is useful for various recovery
+scenarios.
+
+The `to_remote` output has the following form:
+
+  * `OP_1 to_remote_output_key`
+  * where:
+    * `taproot_nums_point = 0245b18183a06ee58228f07d9716f0f121cd194e4d924b037522503a7160432f15`
+    * `to_remote_output_key = taproot_nums_point + tagged_hash("TapTweak", taproot_nums_point || to_remote_script_root)*G`
+    * `to_remote_script_root = tapscript_root([to_remote_script])`
+    * `to_remote_script` is the remote script:
+        ```
+        <remotepubkey> OP_CHECKSIG
+        1 OP_CHECKSEQUENCEVERIFY OP_DROP
+        ```
+
+This output can be swept by the remote party with the following witness:
+```
+<remote_sig> <to_remote_script> <to_remote_control_block>
+```
+
+where `to_remote_control_block` is:
+```
+(output_key_y_parity | 0xc0) || combined_funding_key
+```
+
+The `sequence` field of the input MUST also be set to `1`. 
+
+#### Anchor Outputs
+
+For simple taproot channels (`option_simple_taproot`) we'll maintain the same
+form as base segwit channels, but instead will utilize `local_delayedpubkey`
+and the `remotepubkey` rather than the multi-sig keys as that's no longer
+revealed due to musig2.
+
+An anchor output has the following form:
+
+  * `OP_1 anchor_output_key`
+  * where:
+    * `anchor_internal_key = remotepubkey/local_delayedpubkey`
+    * `anchor_output_key = anchor_internal_key + tagged_hash("TapTweak", anchor_internal_key || anchor_script_root)`
+    * `anchor_script_root = tapscript_root([anchor_script])`
+    * `anchor_script`:
+        ```
+        OP_16 OP_CHECKSEQUENCEVERIFY
+        ```
+
+This output can be swept by anyone after 16 blocks with the following witness:
+```
+<anchor_script> <anchor_control_block>
+```
+
+where `anchor_control_block` is:
+```
+(output_key_y_parity | 0xc0) || anchor_internal_key
+```
+
+`nSequence` needs to be set to 16.
+
+### HTLC Scripts & Transactions
```

Comment:
Did you check compatibility with miniscript? The existing (non-taproot) scripts we use diverge from what the miniscript compiler generates, which is very annoying. Since we have an opportunity to slightly change the scripts we use, I would really like them to be miniscript-compatible, which allows importing the miniscript policies for a forced-close channel into any miniscript-compatible wallet (e.g. bitcoin core), which saves one transaction (the 2nd-level transaction) and thus saves on-chain fees.

@sstone is working on this on our side and can help with the miniscript policies.

Reply:
> Did you check compatibility with miniscript? 

IIRC someone did pretty early on. We had a script that was slightly more optimized, but harder to reason about for humans, so we used a simpler variant. Ultimately, all our scripts are still pretty simple.

I think miniscript compatibility is a laudable goal. FWIW we don't really use it much on the `lnd` side, as we didn't really see a use case as we don't use the `bitcoind` wallet, and subjectively there hasn't been much widespread ecosystem uptake yet (eg: I don't think there's a BIP yet?).

IIUC, only recently has miniscript also been updated to express taproot script trees? 

> The existing (non-taproot) scripts we use diverge from what the miniscript compiler generates, which is very annoying. 

Re the import case you brought up above (I'm not very familiar with this tooling), does this mean that if a script isn't identical to one that would've originally been generated by miniscript, that it can't be import into the `bitcoind` wallet? I thought maybe there'd be some lower level representation (just the script with some additional sugar) that could be used.

Reply:
Re the import case: yeah you need the Miniscript descriptor to precisely describe the script, otherwise you won't find the corresponding coins. I think there may be some confusion with the "Miniscript policy" language. You can import a Miniscript descriptor which has the same semantics as, say, an HTLC output script but you would only find the coins associated with it if you import the script corresponding to this precise implementation of the HTLC semantic.

Reply:
> Re the import case: yeah you need the Miniscript descriptor to precisely describe the script, otherwise you won't find the corresponding coins. I think there may be some confusion with the "Miniscript policy" language. You can import a Miniscript descriptor which has the same semantics as, say, an HTLC output script but you would only find the coins associated with it if you import the script corresponding to this precise implementation of the HTLC semantic.

With the current LN scripts, all we can do is import raw scripts/addresses: bitcoin core won't be aware of any CSV delays and won't know how to spend them.
But our understanding is that, with the ongoing work on descriptors/miniscript/tapscript, if our LN scripts are written in miniscript then every time a channel is force-closed and a commit tx published, we could import a new descriptor that describes our outputs in the commit tx (at least our `to-local` or `to-remote` output) and bitcoin core would know when these outputs become spendable and how to spend them. Once these outputs have been spent then these descriptors could be remove from the wallet (though I believe that is not supported yet).